src/modules/Infil_All.hpp

```diff
@@ -0,0 +1,191 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Canadian Hydrological Model.  If not, see
+// <http://www.gnu.org/licenses/>.
+//
+
+#pragma once
+
+#include "logger.hpp"
+#include "triangulation.hpp"
+#include "module_base.hpp"
+#include "TPSpline.hpp"
+#include <cmath>
+
+
+/**
+ * \ingroup modules infil soils exp
+ * @{
+ * \class Infil_All
+ *
+ * TODO some of this is correct, but likely will be outdated eventually.
+ * Estimates areal snowmelt infiltration into frozen soils for:
+ *    a) Restricted -  Water entry impeded by surface conditions
+ *    b) Limited - Capiliary flow dominates and water flow influenced by soil physical properties
+ *    c) Unlimited - Gravity flow dominates
+ *
+ * **Depends:**
+ * - Snow water equivalent "swe" [mm]
+ * - Snow melt for interval "snowmelt_int" [\f$mm \cdot dt^{-1}\f$]
+ *
+ * **Provides:**
+ * - Infiltration "inf" [\f$mm \cdot dt^{-1}\f$]
+ * - Total infiltration "total_inf" [mm]
+ * - Total infiltration excess "total_excess" [mm]
+ * - Total runoff "runoff" [mm]
+ * - Total soil storage "soil_storage"
+ * - Potential infiltration "potential_inf"
+ * - Opportunity time for infiltration to occur "opportunity_time"
+ * - Available storage for water of the soil "available_storage"
+ *
+ * \rst
+ * .. note::
+ *    Has hardcoded soil parameters that need to be read from the mesh parameters.
+ *
+ * \endrst
+ *
+ * **References:**
+ * - Gray, D., Toth, B., Zhao, L., Pomeroy, J., Granger, R. (2001). Estimating areal snowmelt infiltration into frozen soils
+ * Hydrological Processes  15(16), 3095-3111. https://dx.doi.org/10.1002/hyp.320
+ * @}
+ */
+class Infil_All : public module_base
+{
+REGISTER_MODULE_HPP(Infil_All)
+public:
+    Infil_All(config_file cfg);
+
+    ~Infil_All();
+
+    void run(mesh_elem &face);
+    void init(mesh& domain);
+
+    class data : public face_info
+    {
+    public:
+        struct tempvars
+        {
+            double intensity;
+            double soil_storage_deficit;
+            double capillary_suction;
+            double initial_rate;
+            double initial_storage;
+            double final_rate;
+            double final_storage;
+            double pond;
+            double storage_at_ponding;
+            double time_to_ponding;
+        };
+        double total_inf;
+        double total_excess;
+        double total_snowinf;
+        double total_meltexcess;
+        double total_rain_on_snow;
+        
+        // Crack
+        bool frozen;
+        double index;
+        double max_major_per_melt;
+        double init_SWE;
+        unsigned int major_melt_count;
+        
+        // GreenAmpt
+        double soil_storage;
+        std::unique_ptr<tempvars> GA_temp{nullptr};
```

Comment:
So the idea for this GA_temp variable was to simplify function inputs. It will be accessible as d.GA_temp where the members are tempvars struct.

I could have just had all the members of tempvars be public members of data, but the idea was that these were variables that were not needed between runs (like the other members of data). So I can allocate GA_temp at the start and release it at the end of each call to run ONLY if greenampt is used. This reduces the memory overhead of the module while other modules are running or if greenampt is not used.